staff/lab/timecard

```diff
@@ -0,0 +1,105 @@
+
+    hours = ['{}{}'.format(adjusted_hour(hour), pm_or_am(hour)) for hour in range(24)]
+    sums = [sum(usage[h]) for h in range(24)]
+
+    fig = plt.figure()
+    fig.set_size_inches(12, 4)
+
+    plt.bar(range(24), sums, color='b')
+    plt.xlim(0, 24)
+    plt.xticks(range(24), hours, fontsize=8)
+    plt.xlabel('Hour')
+    plt.ylabel('# Times Present')
+    plt.ylim(0, max(sums) * 1.15)
+
+    plt.title('OCF Lab Timecard for {} ({}-{})'.format(
+        ', '.join(users),
+        start.strftime('%Y-%m-%d'),
+        end.strftime('%Y-%m-%d')
+    ))
+    plt.savefig(dest, bbox_inches='tight')
+
+
+def get_usage_count(*users, start=None, end=None):
+    sem_start, sem_end = stats.semester_dates()
+    if start is None:
+        start = datetime.datetime.combine(sem_start, datetime.time())
+    if end is None:
+        end = datetime.datetime.combine(sem_end, datetime.time())
+
+    step = datetime.timedelta(minutes=60)
+    hours = collections.defaultdict(lambda: [0 for _ in range(7)])
+
+    cnx = stats.get_connection()
+    cursor = cnx.cursor()
+
+    query = """
+        SELECT `start`, `end`
+        FROM `staff_session_duration_public`
+        WHERE `user` IN %s
+            AND `start` >= %s
+            AND `end` <= %s
+            AND `duration` IS NOT NULL
+        ORDER BY start ASC"""
+    cursor.execute(query, (users, start, end))
+    logins = cursor.fetchall()
+
+    was_logged_in = dict()
+    for session in logins:
+        start_time = session['start'].replace(microsecond=0, second=0, minute=0)
+        end_time = session['end']
+        while start_time <= end_time:
+            was_logged_in[start_time] = True
+            start_time += step
+
+    time = start
+    while time <= end:
+        if was_logged_in.get(time):
+            hours[time.hour][time.weekday()] += 1
+        time += step
+
+    return hours, start, end
+
+
+def valid_date(s):
+    # https://stackoverflow.com/a/25470943/6587177
+    try:
+        return datetime.datetime.strptime(s, '%Y-%m-%d')
+    except ValueError:
+        msg = "Not a valid date: '{0}'. Use an isoformat string (YYYY-mm-dd)".format(s)
+        raise argparse.ArgumentTypeError(msg)
+
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser(description='Generate timecard graphs')
+    parser.add_argument('users', type=str, nargs='+', help='User to generate graph for')
+    parser.add_argument('dest', type=str, nargs='?', default='',
+                        help='File to save graph in')
+    parser.add_argument('--start', type=valid_date, default=None,
+                        help='Start date for chart (in form mm/dd/yyyy; defaults to semester start')
+    parser.add_argument('--end', type=valid_date, default=None,
+                        help='Start date for chart (in form mm/dd/yyyy; defaults to semester start')
+    args = parser.parse_args()
+    dest = args.dest if args.dest else '{}.png'.format('_'.join(args.users)[:251])
```

Comment:
Where does `251` come from? Let's not use magic constants like this.

Reply:
Ah, so it seems like the unix filename length is 255 characters, so I strip the prefix name to `251 + 4` from `.png` for a total of `255`. Should a comment be added for that?

Reply:
Is there a built-in constant you can refer to?

Reply:
I can't seem to find any from some google searches. It seems like it's platform dependent and on some OS's depends on the location of the file. Ok to just add a comment explaining where 251 comes from?